### PR TITLE
Use vip dev-env configuration file from parent directories

### PR DESCRIPTION
**Pull request by alecgeatches**
## Description

Extends configuration file support to work in any subdirectory of a VIP site.

`vip dev-env` commands run anywhere within a configured site will automatically use the site's configuration file. `vip dev-env create` will pull environment settings from the `.wpvip` configuration folder, and subsequent `vip dev-env` commands will use the site's configured slug. There's no longer any need to remember a slug or where configuration is setup as long as it's in the root of a VIP site.

In the demo below, a skeleton site is set up with this structure:

```
vip-go-skeleton/
├── .wpvip/
│   └── vip-dev-env.yml      <-- Configuration file here
├── client-mu-plugins/
├── ...
├── themes/
│   └── twentytwentythree/   <-- CLI commands run here
├── private/
└── vip-config/
```

The video demonstrates using the same site's development environment in the site root, and within a theme folder:

https://github.com/Automattic/vip-cli/assets/17870752/2e4f9623-8b1c-46fb-9772-8872408ddbd3

## Additional changes

1. The `.vip-dev-env.yml` file previously could be located anywhere. For consistency, the file is now always loaded from the configuration directory in `.wpvip/vip-dev-env.yml`. Because the configuration directory is hidden, the `.` has been removed from the `vip-dev-env.yml` filename.
1. Changed the supported `configuration-version` from `0.preview-unstable` to `1`.
2. Removed echoing the whole configuration file on `vip dev-env create`.

### Manually

These changes can be tested globally via `npm run build && npm link` using the `add/configuration-file-from-ancestor` branch, or commands can be run directly from compiled files like this:

```bash
$ cd vip-cli
$ npm run build
$ node dist/bin/vip-dev-env-start.js
```

1. Clone [vip-go-skeleton](https://github.com/Automattic/vip-go-skeleton) into a new directory:

    ```bash
    $ git clone git@github.com:Automattic/vip-go-skeleton.git
    ```

3. In the `vip-go-skeleton/` root directory, create a `.wpvip/` folder containing a `vip-dev-env.yml` configuration file:

    ```bash
    $ cd vip-go-skeleton
    $ mkdir .wpvip
    $ vim .wpvip/vip-dev-env.yml
    ```

4. Add the following configuration file contents:

    ```yml
    configuration-version: 1
    slug: test-ancestor-configuration
    title: Test ancestor configuration
    php: 8.0
    wordpress: 6.2
    app-code: ../
    mu-plugins: image
    xdebug: false
    mailpit: false
    multisite: false
    elasticsearch: false
    phpmyadmin: false
    photon: false
    ```

5. In the same root `vip-go-skeleton/` directory, create and run the environment:

    ```bash
    $ vip dev-env create
    $ vip dev-env start
    Using environment test-ancestor-configuration from /Users/alec/.../vip-go-skeleton/.wpvip/vip-dev-env.yml

    # ...
    SLUG              test-ancestor-configuration                                                                                          
    LOCATION          /Users/alec/.local/share/vip/dev-environment/test-ancestor-configuration                                             
    # ...                                                                                                     
    ```

    `vip dev-env create`/`vip dev-env start` commands can also be run in any subdirectory of the configured skeleton site.

6. Verify that `vip dev-env` commands work in subdirectories of `vip-go-skeleton/`

    ```bash
    $ cd plugins/
    $ vip dev-env info
    Using environment test-ancestor-configuration from /Users/alec/.../vip-go-skeleton/.wpvip/vip-dev-env.yml
    # ...
    SLUG              test-ancestor-configuration                                                                                          
    LOCATION          /Users/alec/.local/share/vip/dev-environment/test-ancestor-configuration                                             
    # ...  

    $ vip dev-env exec -- wp option get siteurl
    Using environment test-ancestor-configuration from /Users/alec/.../vip-go-skeleton/.wpvip/vip-dev-env.yml

    Running validation steps...
    # ...
    http://test-ancestor-configuration.vipdev.lndo.site
    ```

7. Try entering deeper subdirectories (e.g. something in `themes/`) and ensure `vip dev-env` commands automatically pull the environment slug from the root site directory.

### Tests

New end-to-end tests have been added in `__tests/devenv-e2e/013-configuration-file.spec.js`. Run these tests with:

```bash
$ npm run test:e2e:dev-env 013
```

### CODE REVIEW

File: src/lib/dev-environment/types.ts [hunk 1]
```diff
@@ -74,9 +74,14 @@ export interface ConfigurationFileOptions {
 	'media-redirect-domain'?: string;
 	photon?: boolean;
 
+	meta?: ConfigurationFileMeta;
```
Comment (alecgeatches) [inline src/lib/dev-environment/types.ts:77]: Added this `meta` key for `ConfigurationFileOptions` to store information about the configuration file that doesn't affect the instance. This meta currently stores `configuration-path` which is used to show this configuration status line:

configuration-file-meta

**Comment by alecgeatches** [comment]
Note: I see there are some type issues causing test failures, fixing those now.

**Comment by rinatkhaziev** [comment]
Just a note on `mailhog` vs `mailpit`, since we ditched the former probably need to update all instances. Step 3: mentions `mailhog`

**Comment by alecgeatches** [comment]
> Just a note on `mailhog` vs `mailpit`, since we ditched the former probably need to update all instances. Step 3: mentions `mailhog`

Thank you! The code appears to be updated to work with `mailpit`/fallback to `mailhog` properly. Updated testing instructions to use `mailpit` instead.

**Comment by alecgeatches** [comment]
The tests have been fixed and this is ready for review.

File: src/lib/constants/dev-environment.ts [hunk 2]
```diff
@@ -28,7 +28,7 @@ export const DEV_ENVIRONMENT_WORDPRESS_CACHE_KEY = 'wordpress-versions.json';
 
 export const DEV_ENVIRONMENT_WORDPRESS_VERSION_TTL = 86400; // once per day
 
-export const DEV_ENVIRONMENT_PHP_VERSIONS: Record< string, string | undefined > = {
+export const DEV_ENVIRONMENT_PHP_VERSIONS: Record< string, string > = {
```
Comment (alecgeatches) [inline src/lib/constants/dev-environment.ts]: Note: Added this change in https://github.com/Automattic/vip-cli/pull/1468/commits/cfb82295b9339b58c2dc1ff5e2aafd59d57a4edd to address an [unrelated type error from tests](https://github.com/Automattic/vip-cli/actions/runs/5744245772/job/15570200021):

```
Run npm run 'check-types'

> @automattic/vip@2.32.4 check-types
> tsc

Error: src/lib/dev-environment/dev-environment-cli.ts(688,4): error TS2322: Type 'string | undefined' is not assignable to type 'string'.
  Type 'undefined' is not assignable to type 'string'.
Error: Process completed with exit code 2.
```

There might be a better way to do this. I'm not sure why `undefined` is part of this type explicitly.

**Comment by alecgeatches** [comment]
Commenting to unstale this. I'm going to re-merge trunk and get this reviewed.

**Comment by rinatkhaziev** [comment]
@alecgeatches I took the liberty and merged the trunk.

**Review by rinatkhaziev — COMMENTED** [review]
Tested, works great. Even went as far as:

```
mkdir -p 1/2/3/4/5/6/7/8/9/10
cd 1/2/3/4/5/6/7/8/9/10 

vip dev-env start
```

On the other hand, this also works:

```
 mv .wpvip ../
 /tmp/vip-go-skeleton   production  vip dev-env start
Using environment test-ancestor-configuration from /private/tmp/.wpvip/vip-dev-env.yml
```
I'm not sure whether this is an imaginary or a real problem, but this may lead to unintended consequences.

**Review by rinatkhaziev — APPROVED** [review]
After further discussion it seems that many other developer tools follow the same traversal strategy, so let's go with it. 

We print the path to the file we use, so it's going to be easier to spot in case the unintended file is being used to start the env.

**Comment by mjangda** [comment]
This is awesome.

Let's make sure to test on Windows as well so that we don't have any unexpected surprises there :)

In a future PR, it could be interesting to support a reference to a VIP environment in the config file for pulling defaults (similar to how you can do `vip @123.develop dev-env create`).

**Comment by rinatkhaziev** [comment]
oh good point re: Windows. Booting up :)

**Comment by rinatkhaziev** [comment]
Tested both under WSL and native Win, works as expected.

**Comment by alecgeatches** [comment]
Note: I made a final change in here to address the security fix made by @sjinks in https://github.com/Automattic/vip-cli/pull/1652 in this commit: https://github.com/Automattic/vip-cli/pull/1468/commits/631f443344c96dea4aba896e139a061774bb307e.

The new code moves `readFile()` into `findConfigurationFile()`, which returns both the configuration path and contents in a `{ configurationPath, configurationContents }` object on success. This should address the same security issue by using a single `readFile()` call to check access and return file contents.

These changes also drop the `ENOENT`/error code checking, as we'll silently drop any errors and continue if a configuration file isn't available.

**Pull request merged**